### PR TITLE
Rewrite BertClassifier using transformers

### LINKED ISSUE #33: Add native HuggingFace Scibert support
In light of https://github.com/allenai/scibert/commit/7598219a8d80b9c2fe1323a141e4a9e40ec044cb, and the closed issue https://github.com/allenai/scibert/issues/71, we don't need cache for Scibert anymore.

It could still be useful to have the "download_and_cache" function if we need another model, but for the time being we can turn it off for Scibert.

### PR DESCRIPTION
**Pull request by nsorros**
Description
---
This PR rewrites BertClassifier using the hugging face transformers library moving away from spacy-transformers. There are a couple of ways we are doing this

- It gives us more flexibility on the features we can use from transformers as spacy-transformers is a wrapper on top that somehow limits what we can do, for example our ability to use all models available.
- Makes it easier to move towards training our own BERT models since a lot of the components can be reused to create a BERT for language modelling
- It decouples slightly from spacy which is generally bad in being backwards compatible, for example upgrading to spacy v3 caused models trained with bert classifier to break even after making the changes necessary to transition

Fixes #33 

Checklist
---

- [x] Added link to Github issue or Trello card
- [x] Added tests

### CODE REVIEW

File: wellcomeml/ml/bert_classifier.py [hunk 1]
```diff
@@ -1,236 +1,191 @@
...
-            config={"exclusive_classes": False, "architecture": "sigmoid_last_hidden"},
-        )
-
-        self.nlp.add_pipe(self.textcat, last=True)
+            return input_data
 
-        for label in self.unique_labels:
-            self.textcat.add_label(label)
+    def _prepare_data(self, X, Y=None, shuffle=False, repeat=False):
+        X_vec = self.tokenizer.batch_encode_plus(
```
Comment (aCampello) [inline wellcomeml/ml/bert_classifier.py:87]: If I am not mistaken, this was the bottleneck for my processing. If using in a CPU, it doesn't parallelise. I am not sure if it parallelise with multi-GPU either. Wonder if the new version of transformers/torch has fixed this.

File: wellcomeml/ml/bert_classifier.py [hunk 2]
```diff
@@ -1,236 +1,191 @@
...
+            dataset = tf.data.Dataset.from_tensor_slices(
+                (X_vec["input_ids"], X_vec["attention_mask"], Y)
+            )
+        else:
+            dataset = tf.data.Dataset.from_tensor_slices(
+                (X_vec["input_ids"], X_vec["attention_mask"])
+            )
+        dataset = dataset.map(self._transform_data)
+        if shuffle:
+            dataset = dataset.shuffle(100, seed=self.random_seed)
```
Comment (aCampello) [inline wellcomeml/ml/bert_classifier.py:102]: Not sure if I understand the variable `buffer_size`, but why do you set to a hard number, 100?

Comment (aCampello) [reply wellcomeml/ml/bert_classifier.py:102]: To be honest, I also do something weird with `buffer_size` :

https://github.com/wellcometrust/WellcomeML/blob/362635afe40779018183aa535b92f5db7d673624/wellcomeml/ml/bert_semantic_equivalence.py#L178

File: wellcomeml/ml/bert_classifier.py [hunk 3]
```diff
@@ -1,236 +1,191 @@
...
+            train_data, validation_data=val_data, epochs=self.epochs,
+            steps_per_epoch=steps_per_epoch, validation_steps=val_steps_per_epoch)
+        self.losses = history.history["loss"]
         return self
 
+    def _predict_proba(self, X):
+        dataset = self._prepare_data(X)
+        out = self.model.predict(dataset)[0]
+        if self.multilabel:
+            return tf.nn.sigmoid(out).numpy()
```
Comment (aCampello) [inline wellcomeml/ml/bert_classifier.py:148]: I see. I do ```tf.keras.activations.sigmoid/softmax```, but I think the functions just call one another on TensorFlow.

**Review by aCampello — APPROVED** [review]
I think the classifier looks good. It looks similar to `bert_semantic_equivalence.py`, with the exception of the data transformation. But even there, I think we just use different verbs, but the data structure is the same. 

It would be good to understand the `buffer_size` parameter of the data transformation function too. After this PR is merged, I will probably improve semantic similarity with the fixed batch size and that.

**Review by aCampello — APPROVED** [review]
Some very minor comments about random seed and shuffle, but other than that looks good.

File: wellcomeml/ml/bert_classifier.py (hunk 1 above)
Comment (nsorros) [reply wellcomeml/ml/bert_classifier.py:87]: Good point, actually my next task is to investigate a bit the tokenisation in order to build a class that we can create a custom tokenizer. So maybe this will give us a way

File: wellcomeml/ml/bert_classifier.py (hunk 2 above)
Comment (nsorros) [reply wellcomeml/ml/bert_classifier.py:102]: yeah i noticed. instead of loading all elements into memory, it takes batches of 100 and shuffles

File: wellcomeml/ml/bert_classifier.py [hunk 4]
```diff
@@ -1,236 +1,191 @@
...
-            path = check_cache_and_download("scibert_scivocab_uncased")
-
-            nlp = TransformersLanguage(trf_name=name, meta={"lang": "en"})
-            nlp.add_pipe(nlp.create_pipe("sentencizer"))
-            nlp.add_pipe(TransformersWordPiecer.from_pretrained(nlp.vocab, path))
-            nlp.add_pipe(TransformersTok2Vec.from_pretrained(nlp.vocab, path))
-            self.nlp = nlp
+        self.max_length = max_length
+        self.multilabel = multilabel
+        self.random_seed = 42
```
[the remark this answers is not in the document]
Comment (nsorros) [reply wellcomeml/ml/bert_classifier.py:58]: 👍

[the remark this answers is not in the document]
Comment (nsorros) [reply wellcomeml/ml/bert_classifier.py]: 0c2dccaac45bdec2cc49c465a41f03002f372a1c

File: wellcomeml/ml/bert_classifier.py (hunk 3 above)
Comment (nsorros) [reply wellcomeml/ml/bert_classifier.py:148]: agree

**Comment by nsorros** [comment]
[Rewrite BertClassifier using transformers and tensorflow](https://trello.com/c/E2wlr2l2/1161-rewrite-bertclassifier-using-transformers-and-tensorflow)
